homeassistant/components/lcn/websocket.py

```diff
@@ -0,0 +1,429 @@
+    async_add_entities([entity])
+
+    # Add entity config to config_entry
+    entity_configs = [*config_entry.data[CONF_ENTITIES], entity_config]
+    data = {**config_entry.data, CONF_ENTITIES: entity_configs}
+
+    # schedule config_entry for save
+    hass.config_entries.async_update_entry(config_entry, data=data)
+
+    connection.send_result(msg["id"], True)
+
+
+@websocket_api.require_admin
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "lcn/entities/delete",
+        vol.Required("host_id"): cv.string,
+        vol.Required(CONF_ADDRESS): ADDRESS_SCHEMA,
+        vol.Required(CONF_DOMAIN): cv.string,
+        vol.Required(CONF_RESOURCE): cv.string,
+    }
+)
+@websocket_api.async_response
+async def websocket_delete_entity(
+    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+) -> None:
+    """Delete an entity."""
+    if not (config_entry := hass.config_entries.async_get_entry(msg["host_id"])):
+        connection.send_result(msg["id"], False)
+        return
+
+    entity_config = next(
+        (
+            entity_config
+            for entity_config in config_entry.data[CONF_ENTITIES]
+            if (
+                tuple(entity_config[CONF_ADDRESS]) == msg[CONF_ADDRESS]
+                and entity_config[CONF_DOMAIN] == msg[CONF_DOMAIN]
+                and entity_config[CONF_RESOURCE] == msg[CONF_RESOURCE]
+            )
+        ),
+        None,
+    )
+
+    if entity_config is None:
+        connection.send_result(msg["id"], False)
+        return
+
+    entity_configs = [
+        ec for ec in config_entry.data[CONF_ENTITIES] if ec != entity_config
+    ]
+    data = {**config_entry.data, CONF_ENTITIES: entity_configs}
+    hass.config_entries.async_update_entry(config_entry, data=data)
+
+    # cleanup registries
+    purge_entity_registry(hass, config_entry.entry_id, data)
+    purge_device_registry(hass, config_entry.entry_id, data)
+
+    connection.send_result(msg["id"])
+
+
+async def async_create_or_update_device_in_config_entry(
+    hass: HomeAssistant,
+    device_connection: DeviceConnectionType,
+    config_entry: ConfigEntry,
+) -> None:
+    """Create or update device in config_entry according to given device_connection."""
+    address = (
+        device_connection.seg_id,
+        device_connection.addr_id,
+        device_connection.is_group,
+    )
+
+    device_configs = [*config_entry.data[CONF_DEVICES]]
+    data = {**config_entry.data, CONF_DEVICES: device_configs}
+    for device_config in data[CONF_DEVICES]:
+        if tuple(device_config[CONF_ADDRESS]) == address:
+            break  # device already in config_entry
+    else:
+        # create new device_entry
+        device_config = {
+            CONF_ADDRESS: address,
+            CONF_NAME: "",
+            CONF_HARDWARE_SERIAL: -1,
+            CONF_SOFTWARE_SERIAL: -1,
+            CONF_HARDWARE_TYPE: -1,
+        }
+        data[CONF_DEVICES].append(device_config)
```

Comment:
Can we cover this case in tests too?

Reply:
I added a special case to the tests, so that at least the code gets covered.